### PR TITLE
Add service active check to weak connections

### PR DESCRIPTION
**Pull request by vicancy**
The service returns `active` ping message when there is any client connected to it, and returns `inactive` ping message when there is no client connected to it. 

For weak connections, there is no need to send server->client messages to a `inactive` service.

From SDK side, we check service status every 10 seconds, whenever we get an `active` ping message, we consider the service as `active`. When we get at least 5 continuous `inactive` ping messages in at least 10 minutes, we consider the service as `inactive`.

Due to the delay of the `active`/`inactive` status, we only enable this for weak connections.

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/WeakServiceConnectionContainer.cs [hunk 1]
```diff
@@ -9,17 +12,189 @@ namespace Microsoft.Azure.SignalR.Common.ServiceConnections
...
+
+                    // Inactive it only when it checks over 5 times and elapsed for over 10 minutes
+                    if (_inactiveCount >= checkWindow && DateTime.UtcNow - _firstInactiveTime >= checkTimeSpan)
+                    {
+                        return false;
+                    }
+                    else
+                    {
+                        return true;
+                    }
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/WeakServiceConnectionContainer.cs]: Add log?

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/ServiceConnections/WeakServiceConnectionContainer.cs]: logged in L44

File: src/Microsoft.Azure.SignalR.Common/ServiceMessages/ServicePingMessageExtension.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+                var item1 = pingMessage.Messages[index];
+                var item2 = pingMessage.Messages[index + 1];
+                if (item1 == key)
+                {
+                    value = item2;
+                    return true;
+                }
+
+                index += 2;
+            }
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Common/ServiceMessages/ServicePingMessageExtension.cs]: would normal ping mixed together, then the order not work?

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/WeakServiceConnectionContainer.cs [hunk 3]
```diff
@@ -9,17 +12,189 @@ namespace Microsoft.Azure.SignalR.Common.ServiceConnections
...
+
+            public static void EndpointInactive(ILogger logger, HubServiceEndpoint endpoint)
+            {
+                _endpointInactive(logger, endpoint, endpoint.Hub, null);
+            }
+
+            public static void EndpointActive(ILogger logger, HubServiceEndpoint endpoint)
+            {
+                _endpointActive(logger, endpoint, endpoint.Hub, null);
+            }
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/WeakServiceConnectionContainer.cs]: logs not use.

File: src/Microsoft.Azure.SignalR.Common/ServiceMessages/ServicePingMessageExtension.cs (hunk 2 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/ServiceMessages/ServicePingMessageExtension.cs]: all the ping follows the same pattern, we use key to distinguish different kind of pings

Comment (JialinXin) [reply src/Microsoft.Azure.SignalR.Common/ServiceMessages/ServicePingMessageExtension.cs]: get it.

---
In reply to: [297487067](https://github.com/Azure/azure-signalr/pull/576#discussion_r297487067) [](ancestors = 297487067)

File: src/Microsoft.Azure.SignalR.Common/ServiceConnections/WeakServiceConnectionContainer.cs [hunk 4]
```diff
@@ -9,17 +12,189 @@ namespace Microsoft.Azure.SignalR.Common.ServiceConnections
...
+        private static readonly TimeSpan DefaultGetServiceStatusInterval = TimeSpan.FromSeconds(10);
+        private static readonly long DefaultGetServiceStatusTicks = DefaultGetServiceStatusInterval.Seconds * Stopwatch.Frequency;
+
+        private readonly TimeSpan CheckTimeSpan = TimeSpan.FromMinutes(10);
+        private readonly object _lock = new object();
+        private int _inactiveCount;
+        private DateTime? _firstInactiveTime;
+        private long _lastSendTimestamp;
+
+        // active ones are those who client connections connected to the whole endpoint
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Common/ServiceConnections/WeakServiceConnectionContainer.cs]: >who [](start = 33, length = 3)

typo: whose

**Pull request merged**